### PR TITLE
Add --max-size flag to limit size of repositories

### PR DESCRIPTION
**Pull request by mholt**
The --max-size limits the number of bytes that can be stored to a repo. It only takes effect on SaveBlob.

Closes https://github.com/restic/rest-server/issues/65

Can see previous discussion at https://github.com/mholt/rest-server/pull/1

/cc @dotlambda

### CODE REVIEW

**Comment by mholt** [comment]
@fd0 I have refactored to make the changes you recommended over in the other PR. But let's move the discussion here. :)

**Comment by mholt** [comment]
There are a "lot" of changed lines because I made the methods get a `*Server` instead of `Server` copy; I think that's needed for correctness of accessing the repoSize value atomically.

Also, I now account for the size as blobs are deleted, for example when a prune happens.

File: handlers.go [hunk 1]
```diff
@@ -566,9 +608,9 @@ func (s Server) DeleteBlob(w http.ResponseWriter, r *http.Request) {
 	}
 
 	var size int64
-	if s.Prometheus {
+	if s.Prometheus || s.MaxRepoSize > 0 {
 		stat, err := os.Stat(path)
-		if err != nil {
+		if err == nil {
```
Comment (mholt) [inline handlers.go]: Technically unrelated to my PR's purpose, but this fixes what I think was a bug that existed before, and with my PR, would affect the size counting if left unfixed.

**Comment by mholt** [comment]
CI fails because golint is broken right now:

> The command "go get -u github.com/golang/lint/golint" failed and exited with 2 during .

File: handlers.go (hunk 1 above)
Comment (fd0) [reply handlers.go]: Oh, great catch

File: cmd/rest-server/main.go [hunk 2]
```diff
@@ -41,6 +41,7 @@ func init() {
 	flags.BoolVar(&server.Debug, "debug", server.Debug, "output debug messages")
 	flags.StringVar(&server.Listen, "listen", server.Listen, "listen address")
 	flags.StringVar(&server.Log, "log", server.Log, "log HTTP requests in the combined log format")
+	flags.Int64Var(&server.MaxRepoSize, "max-size", server.MaxRepoSize, "the maximum size of the repository in bytes")
```
Comment (fd0) [inline cmd/rest-server/main.go]: Hm, why is the default (the third parameter) set to `server.MaxRepoSize`?

Comment (fd0) [reply cmd/rest-server/main.go]: Ok, fine with me ;)

**Pull request merged**